### ISSUE TITLE
Fail to deepcopy a kernel from a fitted model

### ISSUE THREAD

**Issue by PAC-Bayesian** [SheffieldML/GPy#666]
Hi,

I found it is not always possible to copy a kernel. When we initialize a kernel, the `deepcopy` looks fine:  

```python
import numpy as np
import scipy as sp
import matplotlib.pyplot as plt
import GPy
import copy

X = np.arange(10).reshape((10,1))
Y = np.random.randn(10, 1)
k = GPy.kern.RBF(input_dim=1)

k_copy = copy.deepcopy(k)
print(k_copy)
```
```python
rbf.         |  value  |  constraints  |  priors
variance     |    1.0  |      +ve      |        
lengthscale  |    1.0  |      +ve      | 
```

However, **after we pass it into a `GPy` model as an argument**,  `deepcopy` fails and raise `AttributeError`:
```python
m = GPy.models.GPRegression(X, Y, kernel=k)
print(m.kern)
```

```python
rbf.         |  value  |  constraints  |  priors
variance     |    1.0  |      +ve      |        
lengthscale  |    1.0  |      +ve      |
```

```python
k_copy = copy.deepcopy(m.kern)
```

```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-c766c4395347> in <module>()
----> 1 k_copy = copy.deepcopy(m.kern)

~\Anaconda3\envs\build-environment\lib\copy.py in deepcopy(x, memo, _nil)
    159             copier = getattr(x, "__deepcopy__", None)
    160             if copier:
--> 161                 y = copier(memo)
    162             else:
    163                 reductor = dispatch_table.get(cls)

~\Anaconda3\envs\build-environment\lib\site-packages\paramz\core\pickleable.py in __deepcopy__(self, memo)
     99         memo[id(self)] = s # be sure to break all cycles --> self is already done
    100         import copy
--> 101         s.__setstate__(copy.deepcopy(self.__getstate__(), memo)) # standard copy
    102         return s
    103 

~\Anaconda3\envs\build-environment\lib\copy.py in deepcopy(x, memo, _nil)
    148     copier = _deepcopy_dispatch.get(cls)
    149     if copier:
--> 150         y = copier(x, memo)
    151     else:
    152         try:

~\Anaconda3\envs\build-environment\lib\copy.py in _deepcopy_dict(x, memo, deepcopy)
    238     memo[id(x)] = y
    239     for key, value in x.items():
--> 240         y[deepcopy(key, memo)] = deepcopy(value, memo)
    241     return y
    242 d[dict] = _deepcopy_dict

~\Anaconda3\envs\build-environment\lib\copy.py in deepcopy(x, memo, _nil)
    159             copier = getattr(x, "__deepcopy__", None)
    160             if copier:
--> 161                 y = copier(memo)
    162             else:
    163                 reductor = dispatch_table.get(cls)

~\Anaconda3\envs\build-environment\lib\site-packages\paramz\core\pickleable.py in __deepcopy__(self, memo)
     99         memo[id(self)] = s # be sure to break all cycles --> self is already done
    100         import copy
--> 101         s.__setstate__(copy.deepcopy(self.__getstate__(), memo)) # standard copy
    102         return s
    103 

~\Anaconda3\envs\build-environment\lib\site-packages\paramz\parameterized.py in __setstate__(self, state)
    338     #===========================================================================
    339     def __setstate__(self, state):
--> 340         super(Parameterized, self).__setstate__(state)
    341         self._connect_parameters()
    342         self._connect_fixes()

c:\users\riskneutral\desktop\gpy\GPy\core\parameterization\priorizable.py in __setstate__(self, state)
     16 
     17     def __setstate__(self, state):
---> 18         super(Priorizable, self).__setstate__(state)
     19         #self._index_operations['priors'] = self.priors
     20 

~\Anaconda3\envs\build-environment\lib\site-packages\paramz\core\parameter_core.py in __setstate__(self, state)
    493 
    494     def __setstate__(self, state):
--> 495         super(Parameterizable, self).__setstate__(state)
    496         self.logger = logging.getLogger(self.__class__.__name__)
    497         return self

~\Anaconda3\envs\build-environment\lib\site-packages\paramz\core\indexable.py in __setstate__(self, state)
     54 
     55     def __setstate__(self, state):
---> 56         super(Indexable, self).__setstate__(state)
     57         # old index operations:
     58         if not hasattr(self, "_index_operations"):

~\Anaconda3\envs\build-environment\lib\site-packages\paramz\core\pickleable.py in __setstate__(self, state)
    123         from .lists_and_dicts import ObserverList
    124         from ..caching import FunctionCache
--> 125         self.observers = ObserverList()
    126         self.cache = FunctionCache()
    127         self._setup_observers()

~\Anaconda3\envs\build-environment\lib\site-packages\paramz\parameterized.py in __setattr__(self, name, val)
    328         # override the default behaviour, if setting a param, so broadcasting can by used
    329         if hasattr(self, "parameters"):
--> 330             pnames = self.parameter_names(False, adjust_for_printing=True, recursive=False)
    331             if name in pnames:
    332                 param = self.parameters[pnames.index(name)]

~\Anaconda3\envs\build-environment\lib\site-packages\paramz\core\parameter_core.py in parameter_names(self, add_self, adjust_for_printing, recursive, intermediate)
    189         names = []
    190         if intermediate or (not recursive):
--> 191             names.extend([adjust(x.name) for x in self.parameters])
    192         if intermediate or recursive: names.extend([
    193                 xi for x in self.parameters for xi in

~\Anaconda3\envs\build-environment\lib\site-packages\paramz\core\parameter_core.py in <listcomp>(.0)
    189         names = []
    190         if intermediate or (not recursive):
--> 191             names.extend([adjust(x.name) for x in self.parameters])
    192         if intermediate or recursive: names.extend([
    193                 xi for x in self.parameters for xi in

~\Anaconda3\envs\build-environment\lib\site-packages\paramz\core\nameable.py in name(self)
     61         The name of this object
     62         """
---> 63         return self._name
     64     @name.setter
     65     def name(self, name):

AttributeError: 'RBF' object has no attribute '_name'
```
```python
print(m.kern._name)
```
```python
rbf
```

**Comment by PAC-Bayesian** [comment]
It seems I should use the `copy` method defined in the class `paramz.core.pickleable.Pickleable`. Therefore, the following statement works perfect:
```python
k_copy = m.kern.copy()
```
```python
rbf.         |  value  |  constraints  |  priors
variance     |    1.0  |      +ve      |        
lengthscale  |    1.0  |      +ve      |        
```

Sorry for the distraction. Please close this issue!

**Comment by mzwiessele** [comment]
Thanks!

**Comment by AAELsharkawy** [comment]
I am having the same problem when I try to broadcast an object using MPI.bcast() (from mpi4py library) in order to parallelize the code. Is there a way to force the correct deepcopy automatically ?

**Comment by vnmabus** [comment]
I had a similar error three years ago (#583), that was closed without a real resolution. Maybe you have more luck creating a PR that implements `__deepcopy__` for GPy objects, calling the `copy` method, and hoping for the maintainers to care enough to merge it.

In my old problem I had to create a wrapper for `GPy` objects essentially fordwarding every method and attribute access but `__deepcopy__`. However that is a hack and fixing the upstream repository would be better.
